nullaway/src/main/java/com/uber/nullaway/dataflow/AccessPathNullnessPropagation.java

```diff
@@ -530,6 +534,94 @@ private void handleEqualityComparison(
+   *
+   * <ol>
+   *   <li>{@code iter#numX = m.keySet().iterator()}, for getting the iterator over a key set for an
+   *       enhanced-for loop. After such assignments, we track an access path indicating that {@code
+   *       m.get(contentsOf(iter#numX)} is non-null.
+   *   <li>{@code k = iter#numX.next()}, which gets the next key in the key set when {@code
+   *       iter#numX} was assigned as in case 1. After such assignments, we track the desired {@code
+   *       m.get(k)} access path.
+   * </ol>
+   */
+  private void handleEnhancedForOverKeySet(
+      LocalVariableNode lhs,
+      Node rhs,
+      TransferInput<Nullness, NullnessStore> input,
+      ReadableUpdates updates) {
+    if (isEnhancedForIteratorVariable(lhs)) {
+      // Based on the structure of Checker Framework CFGs, rhs must be a call of the form
+      // e.iterator().  We check if e is a call to keySet() on a Map, and if so, propagate
+      // NONNULL for an access path for e.get(iteratorContents(lhs))
+      MethodInvocationNode rhsInv = (MethodInvocationNode) rhs;
+      Node mapNode = getMapNodeForKeySetIteratorCall(rhsInv);
+      if (mapNode != null) {
+        AccessPath mapWithIteratorContentsKey =
+            AccessPath.mapWithIteratorContentsKey(mapNode, lhs, apContext);
+        if (mapWithIteratorContentsKey != null) {
+          updates.set(mapWithIteratorContentsKey, NONNULL);
+        }
+      }
+    } else {
+      // Check for an assignment lhs = iter#numX.next().  From the structure of Checker Framework
+      // CFGs, we know that if iter#numX is the receiver of a call on the rhs of an assignment, it
+      // must be a call to next().
+      if (rhs instanceof MethodInvocationNode) {
+        MethodInvocationNode methodInv = (MethodInvocationNode) rhs;
+        Node receiver = methodInv.getTarget().getReceiver();
+        if (receiver instanceof LocalVariableNode
+            && isEnhancedForIteratorVariable((LocalVariableNode) receiver)) {
+          // See if we are tracking an access path e.get(iteratorContents(receiver)).  If so, since
+          // lhs is being assigned from the iterator contents, propagate NONNULL for an access path
+          // e.get(lhs)
+          AccessPath mapGetPath =
+              input
+                  .getRegularStore()
+                  .getMapGetIteratorContentsAccessPath((LocalVariableNode) receiver);
+          if (mapGetPath != null) {
+            updates.set(AccessPath.withMapKey(mapGetPath, AccessPath.fromLocal(lhs)), NONNULL);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * {@code invocationNode} must represent a call of the form {@code e.iterator()}. If {@code e} is
+   * of the form {@code e'.keySet()}, returns the {@code Node} for {@code e'}. Otherwise, returns
+   * {@code null}.
+   */
+  @Nullable
+  private Node getMapNodeForKeySetIteratorCall(MethodInvocationNode invocationNode) {
```

Comment:
Worth a preconditions check that the method being called is `Set::iterator()`? Or do worry about the extra performance hit of those checks and feel the `isEnhancedForIteratorVariable(...)` check on the caller and our assumptions about CF's CFG generation are enough of a check?

Reply:
In general, I was trying to minimize the overhead from unnecessary checks.  I am very confident in the invariants we rely on in terms of the CF CFG right now.  My hope was that if we bump the CF version and something changes, our tests would fail.  I guess there is the possibility that CF would change in a way that would cause us to start thinking other types of expressions were non-null erroneously, introducing a new unsoundness.  That seems pretty unlikely to me, but if you'd like, I can introduce some extra sanity checking.  I would want to limit that checking to not run for every assignment statement, though, instead only for cases where we are pretty sure we're in an enhanced for loop. WDYT?

Reply:
Your call. I think you are way more familiar with the CF internals than I am, if you feel the invariant is pretty much set in stone and requires no runtime checking, I am fine with that. That said, if you think there is a reasonable value in adding sanity checking, I definitely agree we can defer it until right before the `updates.set(...)` call. Though, again, if you think our assumptions are unlikely to break, I am also fine leaving this as is.